### PR TITLE
Documentation generation updates

### PR DESCRIPTION
**Pull request by phiggins**
To view these docs:
* check out this branch
* run `make serve` from the `automate-chef-io` directory
* visit `http://localhost:1313/docs/api/`

To make modifications:
* check out this branch
* run `make serve` from the `automate-chef-io` directory (when left running this command watches for changes to files it knows about and will rebuild the hugo site automatically)
* run `compile_all_protobuf_components` in a hab studio
* run `make sync_swagger_files generate_swagger`from the `automate-chef-io` directory
* visit `http://localhost:1313/docs/api/`

Or to compile everything in one copy-pastable command:
```
hab studio run "source .studiorc && compile_all_protobuf_components" && pushd components/automate-chef-io/ && make sync_swagger_files generate_swagger && make serve || popd
```

### CODE REVIEW

File: api/interservice/authz/v2/project.proto [hunk 1]
```diff
@@ -2,7 +2,6 @@ syntax = "proto3";
 
 import "validate/validate.proto";
 import "api/interservice/authz/v2/type.proto";
-import "api/interservice/event/event.proto";
```
Comment (phiggins) [inline api/interservice/authz/v2/project.proto]: This is unrelated, I was just tired of seeing the warning generated during proto compilation from this.

File: components/automate-chef-io/Makefile [hunk 2]
```diff
@@ -3,7 +3,10 @@
 SHELL=bash
 SWAGGER_RESULT_FILE=static/api-docs/all-apis.swagger.json
 SWAGGER_DIR=data/docs/api_chef_automate
-SWAGGER_FILES = $(shell find $(SWAGGER_DIR) -name '*.swagger.json')
+#SWAGGER_FILES=$(shell find $(SWAGGER_DIR) -name '*.swagger.json')
+SWAGGER_FILES=data/docs/api_chef_automate/compliance/reporting/reporting.swagger.json data/docs/api_chef_automate/gateway/gateway.swagger.json
+STATIC_SWAGGER_FILES=$(shell find data/docs/api-static/ -type f | sort -n)
```
Comment (phiggins) [inline components/automate-chef-io/Makefile]: Adding files to `data/docs/api-static/` is how I implemented adding things that can't come from the proto files (eg, reporting server's export endpoint) or would be awkward to put in a specific proto file (eg, the metadata that applies to the docs as a whole). They are sorted numerically so that things happen in a deterministic way.

File: components/automate-gateway/api/compliance/reporting/reporting.proto [hunk 3]
```diff
@@ -31,24 +31,21 @@ service ReportingService {
...
+		};
+		option (chef.automate.api.policy) = {
+			resource: "compliance:reporting:report-ids"
+			action: "search"
+		};
+		option (chef.automate.api.iam.policy) = {
+			resource: "compliance:reporting:report-ids"
+			action: "compliance:reportids:list"
+		};
+	};
```
Comment (phiggins) [inline components/automate-gateway/api/compliance/reporting/reporting.proto]: The diff here is mostly due to whitespace changes.

File: components/automate-gateway/api/compliance/reporting/reporting.proto [hunk 4]
```diff
@@ -323,20 +339,29 @@ message Statistics {
 }
 
 message SuggestionRequest {
+	// The type of resource to get suggestions for
 	string type = 1;
+	// The term to use to match resources on 
 	string text = 2;
+	// The maximum number of suggestions to return
 	int32 size = 3;
+	// The criteria used to filter the suggestions returned.
```
Comment (phiggins) [inline components/automate-gateway/api/compliance/reporting/reporting.proto]: This ends with a `.` to turn it from a `title` to a `description`. For some reason, the `title` of an array will not appear in the ReDoc UI.

File: components/automate-chef-io/layouts/_default/data-api.html [hunk 5]
```diff
@@ -1,7 +1,7 @@
 <!DOCTYPE html>
 <html>
   <head>
-    <title>ReDoc</title>
+    <title>Chef Automate documentation</title>
```
Comment (susanev) [inline components/automate-chef-io/layouts/_default/data-api.html]: should this be `Chef Automate API Documentation`?

File: components/automate-chef-io/Makefile (hunk 2 above)
Comment (srenatus) [reply components/automate-chef-io/Makefile]: 💡 This is a nice idea. I was wondering how to deal with those.

File: components/automate-chef-io/layouts/_default/data-api.html (hunk 5 above)
Comment (phiggins) [reply components/automate-chef-io/layouts/_default/data-api.html]: All of the text in here should be regarded as barely informed placeholders :)

This came from me wondering where the page title came from, initially suspecting it was from metadata, but it turned out to be a plain old html title. I can update this.

**Comment by phiggins** [comment]
> is starting everything with The the standard we want?

If there is a convention we want to follow I'd be happy to make changes.

**Comment by susanev** [comment]
that was mostly a question for @kagarmoe, i should have been more clear. in my experience, the scanning of the content may be easier if we dont start everything with the same word.

File: components/automate-chef-io/Makefile [hunk 6]
```diff
@@ -38,8 +41,13 @@ sync_swagger_files: clean_swagger_files
 
 $(SWAGGER_RESULT_FILE): $(SWAGGER_FILES)
 
+#generate_swagger: $(SWAGGER_RESULT_FILE)
+#	swagger mixin --quiet -c=$(EXPECTED_CONFLICTS) $(SWAGGER_FILES) > $(SWAGGER_RESULT_FILE)
+
 generate_swagger: $(SWAGGER_RESULT_FILE)
-	swagger mixin --quiet -c=$(EXPECTED_CONFLICTS) $(SWAGGER_FILES) > $(SWAGGER_RESULT_FILE)
+	swagger mixin $(STATIC_SWAGGER_FILES) $(SWAGGER_FILES) | \
+		jq '.paths = (.paths | with_entries( select( all( .value[].tags[]; . != "hidden") ) ) )' \
```
Comment (phiggins) [inline components/automate-chef-io/Makefile]: In English, this says to replace the `paths` object in the swagger doc with a new copy that has any endpoint with a tag named "hidden" removed. 

The swagger path to tags looks generally like this:

```
{
  ...
  "paths": {
    "/path/to/my/endpoint": {
       "HTTP_METHOD": {
         ...
         "tags": ["foo", "bar"],
```

Comment (srenatus) [reply components/automate-chef-io/Makefile]: 👏 nicely done

**Comment by phiggins** [comment]
External things I used for research, mostly around inspec things:
https://www.inspec.io/docs/reference/profiles/
https://www.inspec.io/docs/reference/dsl_inspec/
http://www.anniehedgie.com/inspec-basics-9

**Comment by phiggins** [comment]
There is a small bug that I have not been able to find a solution to. Any place where a singular proto message is embedded in another message, we do not get fine grained control over the documentation for that item. An example:

```
message ControlSummary {
	// The total number of controls in the report
	int32 total = 1;
	// Intentionally blank
	Total passed = 2;
	// Intentionally blank
	Total skipped = 3;
	// Intentionally blank
	Failed failed = 4;
}
```

The `passed` and `skipped` entries in `ControlSummary` both pull their metadata from the message definition of `Total` rather than where the individual entries are defined. This means that in this case `passed` and `skipped` will get the same metadata. I've marked all the occurences of this with the comment `// Intentionally blank` for two main reasons:

* It lets folks editing the file know that this is a placeholder that will not show up in the UI
* If something changes, it will hopefully make that more obvious when that text shows up in the UI.

The only reference I could find to this was in the ReDoc bug tracker where it was described as a limitation of Swagger: https://github.com/Redocly/redoc/issues/835

File: components/automate-gateway/api/compliance/reporting/reporting.proto [hunk 7]
```diff
@@ -92,8 +109,15 @@ service ReportingService {
 			resource: "compliance:reporting:profiles"
 			action: "compliance:reportProfiles:list"
 		};
+		option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
+			description: "List all profiles optionally using filters";
```
Comment (vjeffrey) [inline components/automate-gateway/api/compliance/reporting/reporting.proto]: i wonder if there are words we can use here to make it clear that these are profiles as in profile results from a scan..?

File: components/automate-gateway/api/compliance/reporting/reporting.proto [hunk 8]
```diff
@@ -151,247 +194,400 @@ message TimeQuery {
...
 	Platform platform = 9;
+	// Intentionally blank
 	Statistics statistics = 10;
+	// The profiles run as part of this report
 	repeated Profile profiles = 11;
+	// The id of the job associated with the report (???)
 	string job_id = 12;
+	// The IP address of the node generating the report
 	string ipaddress = 13;
+	// The FQDN (fully qualified domain name) of the node generating the report
```
Comment (vjeffrey) [inline components/automate-gateway/api/compliance/reporting/reporting.proto]: mmmm so many comments yeesssssss! 🎉

File: components/automate-gateway/api/compliance/reporting/reporting.proto [hunk 9]
```diff
@@ -151,247 +194,400 @@ message TimeQuery {
...
+	// The short name of the dependency stored on the Chef Automate or Chef Compliance server
 	string compliance = 11;
+	// The status of the dependency in the report
 	string status = 12;
+	// The reason this profile was skipped in the generated report, if any
 	string skip_message = 13;
 }
 
 message Group {
+	// ???
```
Comment (vjeffrey) [inline components/automate-gateway/api/compliance/reporting/reporting.proto]: uh tbh i don't remember what this means either...i am making a note to educate myself.

File: components/automate-gateway/api/compliance/reporting/reporting.proto [hunk 10]
```diff
@@ -151,247 +194,400 @@ message TimeQuery {
...
 }
 
 message Suggestion {
+	// The content that matched the search term
 	string text = 1;
+	// The id of the resource that was suggested
 	string id = 2;
+	// The confidence in the match quality
 	float score = 3;
+	// ???
```
Comment (vjeffrey) [inline components/automate-gateway/api/compliance/reporting/reporting.proto]: ..ya what is version in this? @alexpop ?

**Comment by vjeffrey** [comment]
sounds like we're pretty ready to take this out of draft mode and get it merged in.
**what are the things we absolutely must do before merging this?**

**Comment by kagarmoe** [comment]
We need to adjust the tags, descriptions, and summaries so the endpoints are presented in a meaningful order. I need Pete to show me where in the code to make these adjustments and to help me through getting set up to build and see changes. I will see how much of my time I can clear for this next week.

**Comment by phiggins** [comment]
> We need to adjust the tags, descriptions, and summaries so the endpoints are presented in a meaningful order. 

I don't know how much control we have over that vs everything being in alphabetical order. I'll look in the Redoc docs to see if there are any knobs we can tweak there.

**Comment by kagarmoe** [comment]
I've already found some ways, but I need to know how to make those changes.

**Comment by phiggins** [comment]
> I've already found some ways, but I need to know how to make those changes.

Can you point me to what you're referring to?

**Comment by kagarmoe** [comment]
Questions on the `modifications` steps

For `run compile_all_protobuf_components in a hab studio`, do I need to be in the top-level automate directory? Do I need to run `start_all_services` first?

**Comment by vjeffrey** [comment]
chatted a bit with kimberly on friday. sounds like the following is required in order to merge this pr:
- fill in authentication section: [example](https://clearbit.com/docs?utm_source=zapier.com&utm_medium=referral&utm_campaign=zapier#authentication)
- make some decisions about how to present code samples.
- establish minimum languages for the API call examples

we're also going to meet wednesday to make sure we got everything on the list/figure out post-merge work to be done

**Comment by susanev** [comment]
@vjeffrey does auth need to take on the first bullet?

**Comment by vjeffrey** [comment]
@susanev thanks for asking! i think we can cover this, though we may have a couple questions.
